### ISSUE TITLE
0.4.14-dev - Several servers repeatedly crashing: vector.lua Line 92

### ISSUE THREAD

**Issue by Cat5TV** [minetest/minetest#4132]
Minetest 0.4.14-dev-7a828de (Linux)
Build info: VER=0.4.14-dev-7a828de BUILD_TYPE=Release RUN_IN_PLACE=0 USE_GETTEXT=0 USE_SOUND=0 USE_CURL=1 USE_FREETYPE=1 USE_LUAJIT=0 STATIC_SHAREDIR="/usr/local/share/minetest"

**Error on Server # 1:**
ERROR[Main]: ServerError: Lua: Runtime error from mod 'moretrees' in callback LuaABM::trigger(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)

**Error later on Server # 2:**
ERROR[Main]: ServerError: Lua: Runtime error from mod '*builtin*' in callback node_on_dig() : ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)

**And later still on Server # 1 again:**
ERROR[Main]: ServerError: Lua: Runtime error from mod 'moresnow' in callback luaentity_Step(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)

First thought (first error above) was that moretrees was part of the cause. Then the same error occurred two more times, but seemingly unrelated mods.

Thanks  :)

**Comment by sofar** [comment]
`vector.lua:91` is `vector.add()`

It would be good to know what ABM that was. Does the `moresnow` mod use ABM's?

**Comment by paramat** [comment]
https://github.com/minetest/minetest/blob/master/builtin/common/vector.lua#L92

**Comment by paramat** [comment]
https://github.com/Sokomine/moresnow/blob/master/snow_cannon.lua the only luaentity step i can see in moresnow, vector.add is not used but perhaps it's being called by obj:set velocity/acceleration.

**Comment by PilzAdam** [comment]
Would be nice if we got full backtraces.

**Comment by Cat5TV** [comment]
Sure--anything you folks need, just let me know.

As I look at these full backtraces, I realize this also explains why some users crash the server as soon as they join (another issue I was talking about on IRC today)...

**--- First crash yesterday**

2016-05-15 11:05:51: ACTION[Server]: pkelley817 digs default:jungleleaves at (2493,6,581)
2016-05-15 11:05:52: ACTION[Main]: IRC: Disconnected.
2016-05-15 11:05:54: ERROR[Main]: ServerError: Lua: Runtime error from mod '*builtin*' in callback LuaABM::trigger(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-15 11:05:54: ERROR[Main]: stack traceback:
2016-05-15 11:05:54: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-15 11:05:54: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-15 11:05:54: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-15 11:05:54: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-15 11:05:54: ERROR[Main]: 	...games/thepixelshadow_game/mods/default/functions.lua:349: in function <...games/thepixelshadow_game/mods/default/functions.lua:285>

**--- Second crash: note it's the same player who chopped down the jungleleaves (falling)**

2016-05-15 11:08:40: ACTION[Server]: pkelley817 [76.215.53.68] joins game. 
2016-05-15 11:08:40: ACTION[Server]: pkelley817 joins game. List of players: Vince dkelley810 Batman pkelley817
2016-05-15 11:08:41: ACTION[Main]: IRC: Disconnected.
2016-05-15 11:08:42: ERROR[Main]: ServerError: Lua: Runtime error from mod '*builtin*' in callback LuaABM::trigger(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-15 11:08:42: ERROR[Main]: stack traceback:
2016-05-15 11:08:42: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-15 11:08:42: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-15 11:08:42: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-15 11:08:42: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-15 11:08:42: ERROR[Main]: 	...games/thepixelshadow_game/mods/default/functions.lua:349: in function <...games/thepixelshadow_game/mods/default/functions.lua:285>

**---**

2016-05-15 11:12:37: ACTION[Server]: pkelley817 [76.215.53.68] joins game. 
2016-05-15 11:12:37: ACTION[Server]: pkelley817 joins game. List of players: Batman dkelley810 pkelley817
2016-05-15 11:12:38: ACTION[Main]: IRC: Disconnected.
2016-05-15 11:12:39: ERROR[Main]: ServerError: Lua: Runtime error from mod 'moretrees' in callback LuaABM::trigger(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-15 11:12:39: ERROR[Main]: stack traceback:
2016-05-15 11:12:39: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-15 11:12:39: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-15 11:12:39: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-15 11:12:39: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-15 11:12:39: ERROR[Main]: 	...games/thepixelshadow_game/mods/default/functions.lua:349: in function <...games/thepixelshadow_game/mods/default/functions.lua:285>

**--- Moving along ... I had not made the connection until now. Different player, same result:**

2016-05-15 13:47:38: ACTION[Server]: TSmith digs default:jungleleaves at (1813,36,-133)
2016-05-15 13:47:38: ACTION[Main]: IRC: Disconnected.
2016-05-15 13:47:39: ERROR[Main]: ServerError: Lua: Runtime error from mod 'moretrees' in callback LuaABM::trigger(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-15 13:47:39: ERROR[Main]: stack traceback:
2016-05-15 13:47:39: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-15 13:47:39: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-15 13:47:39: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-15 13:47:39: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-15 13:47:39: ERROR[Main]: 	...games/thepixelshadow_game/mods/default/functions.lua:349: in function <...games/thepixelshadow_game/mods/default/functions.lua:285>
Minetest Stopped. List of players:

**--- Now this time, not leaves, but same player as the first error**

2016-05-16 16:58:42: ACTION[Server]: pkelley817 digs mesecons_solarpanel:solar_panel_on at (2518,2,602)
2016-05-16 16:58:42: ACTION[Main]: IRC: Disconnected.
2016-05-16 16:58:42: ACTION[Server]: MangleFox70 leaves game. List of players: 
2016-05-16 16:58:42: ACTION[Server]: pkelley817 leaves game. List of players: 
2016-05-16 16:58:42: ACTION[Server]: nudistgamer leaves game. List of players: 
2016-05-16 16:58:42: ACTION[Server]: Batman leaves game. List of players: 
2016-05-16 16:58:42: ACTION[Server]: dkelley810 leaves game. List of players: 
2016-05-16 16:58:42: ACTION[Server]: KandyBloodrose leaves game. List of players: 
2016-05-16 16:58:42: ACTION[Server]: Joe leaves game. List of players: 
2016-05-16 16:58:46: ERROR[Main]: ServerError: Lua: Runtime error from mod '*builtin*' in callback node_on_dig(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-16 16:58:46: ERROR[Main]: stack traceback:
2016-05-16 16:58:46: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-16 16:58:46: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-16 16:58:46: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-16 16:58:46: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-16 16:58:46: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:258: in function 'callback'
2016-05-16 16:58:46: ERROR[Main]: 	/home/robbie/minetest/bin/../builtin/game/item.lua:511: in function </home/robbie/minetest/bin/../builtin/game/item.lua:439>
2016-05-16 16:58:46: ERROR[Main]: 	(tail call): ?

**--- Then later today, same error BUT no players were online, nor was anyone signing in**

2016-05-16 17:00:01: ERROR[Main]: ServerError: Lua: Runtime error from mod '*builtin*' in callback luaentity_Step(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-16 17:00:01: ERROR[Main]: stack traceback:
2016-05-16 17:00:01: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-16 17:00:01: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-16 17:00:01: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-16 17:00:01: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-16 17:00:01: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:258: in function 'callback'
2016-05-16 17:00:01: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:92: in function <...home/robbie/minetest/bin/../builtin/game/falling.lua:43>

That happened a few times in a row...

**--- Now the moretrees one I posted earlier. Again, related to falling?**

2016-05-16 17:07:04: ERROR[Main]: ServerError: Lua: Runtime error from mod 'moretrees' in callback LuaABM::trigger(): ...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)
2016-05-16 17:07:04: ERROR[Main]: stack traceback:
2016-05-16 17:07:04: ERROR[Main]: 	...ome/robbie/minetest/bin/../builtin/common/vector.lua:92: in function 'add'
2016-05-16 17:07:04: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:137: in function 'check_attached_node'
2016-05-16 17:07:04: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:171: in function 'nodeupdate_single'
2016-05-16 17:07:04: ERROR[Main]: 	...home/robbie/minetest/bin/../builtin/game/falling.lua:220: in function 'nodeupdate'
2016-05-16 17:07:04: ERROR[Main]: 	...games/thepixelshadow_game/mods/default/functions.lua:349: in function <...games/thepixelshadow_game/mods/default/functions.lua:285>

--- 

That's probably enough since they all tend to run together.  :)

Let me know if you need anything else.

**Comment by sofar** [comment]
Possible far-fetched cause: `check_attached_node()` uses `core.wallmounted_to_dir(n.param2)` if the nodetype is `wallmounted`. If `param2` is not [0-5] but instead 6 or higher, this will return in a `nil` in line `137` of falling.lua causing the crash.

It would be good to `assert()` that wallmounted_to_dir() is not outside the valid range.

@Cat5TV can you run with this patch applied to see if this is the case? :

```
diff --git a/builtin/game/item.lua b/builtin/game/item.lua
index 36c2c1a..e4b7ad6 100644
--- a/builtin/game/item.lua
+++ b/builtin/game/item.lua
@@ -142,6 +142,8 @@ local wallmounted_to_dir = {
        {x =  0, y =  0, z = -1},
 }
 function core.wallmounted_to_dir(wallmounted)
+       assert(wallmounted >= 0)
+       assert(wallmounted < 6)
        return wallmounted_to_dir[wallmounted]
 end
 
```

**Comment by Cat5TV** [comment]
Thanks sofar. How you made that connection is well beyond me. Finger's crossed! I have applied the patch against today's build and will keep an eye on it and post any crashes should they occur.

**Comment by sofar** [comment]
detailed reasoning:

The crash is in `vector.lua:92: attempt to perform arithmetic on local 'b' (a nil value)`

line 92 of vector.lua is `function vector.add(a, b)`, so the *second* parameter is `nil`.

Back to `check_attached_node`:

```
1290 function check_attached_node(p, n)
130	local def = core.registered_nodes[n.name]
131	local d = {x = 0, y = 0, z = 0}
132	if def.paramtype2 == "wallmounted" then
133		d = core.wallmounted_to_dir(n.param2)
134	else
135		d.y = -1
136	end
137	local p2 = vector.add(p, d)
```

Line 137 calls vector.add with `d` as second parameter for `vector.add` and during the crash, `d = nil`.

If we read the code we can see that `d = {x=0, y=0, z=0}` by default. And that normally `d.y = -1` for most nodes. This would make `d` a table with `{x=0, y=-1, z=0}` and this should not be any issues for vector.add, since it would `return` from `vector.add()` at line 88. 

So we know that `type(b)` in `vector.add` is not a `"table"` since it crashes in line 92.

So logically it follows that `def.paramtype2 == "wallmounted"` and that `core.wallmounted_to_dir(n.param2)` was executed.

`core.wallmounted_to_dir()` does a table lookup for `n.param2` in a table that has 6 elements, starting at `0` and ending at `5`. If `n.param2` is in that range, it returns a `table`. If it is outside that range, `nil`.

**Comment by sofar** [comment]
My bet is that some mod is setting `param2 = 6` and assumes that that is identical to `0`.

**Comment by Cat5TV** [comment]
My dear Watson...
Is there something I could grep through the installed mods (eg., "param2 = 6") to see which mod it could be, if that's the case?

**Comment by sofar** [comment]
My suspicion is vines, or biome_lib. It would be good to see if there are other nodes growing against the sides of jungleleaves on your server.

however, I took a deep look at the biome_lib and vines code and can't find any code that would make p2 = 6+ directly.

**Comment by Cat5TV** [comment]
May 29 - getting the server crashing with this issue today.

Anything else I can do to help track this problem down?

**Comment by sofar** [comment]
Did you leave the asserts in place?

**Comment by Cat5TV** [comment]
Hah, no of course I didn't, because I'm dumb like that (had to recompile a few days ago).

Do the asserts *only* tell you stuff, or do they actually stop this issue.

Putting the asserts back now...

**Comment by Cat5TV** [comment]
UPDATE: I have re-applied the asserts (like a boss)

**Comment by sofar** [comment]
What I want to see is at least one assert crashing due to param2 >= 6. Then I'll submit a PR to avoid the crash.

**Comment by paramat** [comment]
See #4557

**Comment by paramat** [comment]
Shall we perhaps add a workaround (limiting param2 value) but keep this open?

**Comment by paramat** [comment]
Copied from Megaf's issue #4557 
/////////////////////////

The way I caused this crash this time was as it follows.
Selected 1x1x1 node inside a water_flowing area, in the air.
Set the node to torch
Dug the torch
Server crashed.

If you select an area of air lets say, 30x30x30, and set everything to torch, it might crash the server too.

```
WorldEdit -!- position 1 set to (194,16,-3423)
<ACTION> 2016-09-28 07:27:03: [Server] buttatoes places node default:obsidian_glass at (196,87,-3423)
WorldEdit -!- position 2 set to (194,16,-3423)
WorldEdit -!- 1 nodes set
WorldEdit -!- region reset
<ACTION> 2016-09-28 07:27:08: [Server] A sapling grows into a tree at (478,21,-588)
<ACTION> 2016-09-28 07:27:10: [Server] buttatoes places node default:obsidian_glass at
(195,87,-3422)
<ACTION> 2016-09-28 07:27:10: [Server] buttatoes places node default:obsidian_glass at
(195,86,-3423)
[biome_lib] Stand by, playing out the rest of the aircheck mapblock log
(there are 0 entries)...
[biome_lib] Stand by, playing out the rest of the no-aircheck mapblock log
(there are 0 entries)...
<ACTION> 2016-09-28 07:27:13: [Server] Megaf digs default:torch at (194,16,-3423)
End log output over terminal (no stdout/stderr backlog during that)
========================
2016-09-28 07:27:13: ERROR[Main]: ServerError: Lua: Runtime error from mod 'default' in callbac
k node_on_punch(): ...afServer/BinMegafServer/bin/../builtin/common/vector.lua:92: attempt to p
erform arithmetic on local 'b' (a nil value)
2016-09-28 07:27:13: ERROR[Main]: stack traceback:
2016-09-28 07:27:13: ERROR[Main]:       ...afServer/BinMegafServer/bin/../builtin/common/vector
.lua:92: in function 'add'
2016-09-28 07:27:13: ERROR[Main]:       ...gafServer/BinMegafServer/bin/../builtin/game/falling
.lua:137: in function 'check_attached_node'
2016-09-28 07:27:13: ERROR[Main]:       ...gafServer/BinMegafServer/bin/../builtin/game/falling
.lua:171: in function 'nodeupdate_single'
2016-09-28 07:27:13: ERROR[Main]:       ...gafServer/BinMegafServer/bin/../builtin/game/falling
.lua:220: in function 'nodeupdate'
2016-09-28 07:27:13: ERROR[Main]:       ...gafServer/BinMegafServer/bin/../builtin/game/falling
.lua:263: in function 'callback'
2016-09-28 07:27:13: ERROR[Main]:       .../MegafServer/BinMegafServer/bin/../builtin/game/item
.lua:419: in function <.../MegafServer/BinMegafServer/bin/../builtin/game/item.lua:412>
```

The server is not a very old git master. The bug is old actually.
And I don't play singleplayer, I'm always on my server.

**Comment by Fixer-007** [comment]
secret minetest development server?

**Comment by t4im** [comment]
This crash indeed happens with wallmount param2 >= 6, which is easily triggered by mods using Voxelmanip (or perhaps even schematics) to place wall mounted stuff in the world, without resetting param2. Worldedit is no exception here, which might be the main cause.

**Comment by paramat** [comment]
Schematics autoset param2 to 0 if not specified (i think).

**Comment by paramat** [comment]
Fix #4565

**Comment by paramat** [comment]
#4565 merged but keeping this open for a short while.
